feat: add UDP in HTTP

imgk opened the pull request:
<!--
Thank you for contributing to Caddy! Please fill this out to help us make the most of your pull request.
-->

### 1. What does this change do, exactly?

Add UDP in HTTP support

### 2. Please link to the relevant issues.

https://github.com/klzgrad/naiveproxy/issues/617

### 3. Which documentation changes (if any) need to be made because of this PR?


### 4. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code

klzgrad commented on forwardproxy.go:
```diff
@@ -240,6 +249,19 @@ func (h *Handler) Provision(ctx caddy.Context) error {
+	h.udpProxyServer = &udpProxyServer{}
+	// parse http2/http3 uri template
+	if h.URITemplate == "" {
+		h.uriTemplate = uritemplate.MustNew("https://{host}/.well-known/masque/udp/{target_host}/{target_port}/")
```
Is it necessary to introduce a library for this? Can you use simpler functions?

imgk replied:
Removed

klzgrad commented on forwardproxy_udp.go:
```diff
@@ -0,0 +1,593 @@
+	return Request(net.JoinHostPort(targetHost, strconv.Itoa(targetPort))), nil
+}
+
+func (h *Handler) checkACL(hostPort string) (bool, error) {
```
Seems to have duplicated forwardproxy.go.

I think much of this should be embedded along side the logic of tcp/http.

imgk replied:
Done

naiba commented:
any update?

klzgrad commented:
Review comments are not addressed yet.

naiba commented:
How is it now? This PR looks very interesting, does the naiveproxy client need to be modified?

klzgrad commented:
I asked for not duplicating forwardproxy.go into a udp version, and there was no answer for that.

imgk commented:
> I asked for not duplicating forwardproxy.go into a udp version, and there was no answer for that.

Hi, I misunderstood your request there. I believe this is done now.

naiba commented:
Hi @imgk, Does this feature require any changes to the client? Or does it already support the client that complies with the RFC definition?

imgk commented:
> Hi @imgk, Does this feature require any changes to the client? Or does it already support the client that complies with the RFC definition? 

Hi, this modified `forwardproxy` should be compatible with any client which supports RFC 9298. However, this module is not well tested yet. The function is not guaranteed. Feedbacks are welcomed.

PrintNow committed:
docs: fix `dial_timeout` parameter documentation (#171)

Fixes #170

- Update `dial_timeout` example to use duration unit (30s)
- Change parameter type from [integer] to [Duration] to match Go type
- Update description to clarify duration unit requirement

This change makes the documentation more accurate and consistent
with Go's time.Duration type usage.

dependabot[bot] committed:
Bump actions/checkout from 4 to 5 (#174)

Bumps [actions/checkout](https://github.com/actions/checkout) from 4 to 5.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/golang/glog from 1.2.0 to 1.2.4 (#175)

Bumps [github.com/golang/glog](https://github.com/golang/glog) from 1.2.0 to 1.2.4.
- [Release notes](https://github.com/golang/glog/releases)
- [Commits](https://github.com/golang/glog/compare/v1.2.0...v1.2.4)

---
updated-dependencies:
- dependency-name: github.com/golang/glog
  dependency-version: 1.2.4
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump actions/setup-go from 5 to 6 (#178)

Bumps [actions/setup-go](https://github.com/actions/setup-go) from 5 to 6.
- [Release notes](https://github.com/actions/setup-go/releases)
- [Commits](https://github.com/actions/setup-go/compare/v5...v6)

---
updated-dependencies:
- dependency-name: actions/setup-go
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

aaravrav commented:
How is the status of the PR now?

dependabot[bot] committed:
Bump github.com/quic-go/quic-go from 0.44.0 to 0.49.1 (#182)

Bumps [github.com/quic-go/quic-go](https://github.com/quic-go/quic-go) from 0.44.0 to 0.49.1.
- [Release notes](https://github.com/quic-go/quic-go/releases)
- [Commits](https://github.com/quic-go/quic-go/compare/v0.44.0...v0.49.1)

---
updated-dependencies:
- dependency-name: github.com/quic-go/quic-go
  dependency-version: 0.49.1
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

imgk committed:
add UDP in HTTP

naiba reviewed (commented):
## Code Review (AI-assisted review by Claude)

Thanks for the implementation! Reviewed against RFC 9298/9297 and the existing codebase. A few issues found:

### Blocking

**1. HTTP/2 path uses `Hijack()` — breaks HTTP/2 multiplexing**

```go
case 2:
    conn, _, err := rc.Hijack()
```

`Hijack()` drops down to the raw TCP connection, bypassing HTTP/2 framing entirely. This destroys all other streams on the same connection. The existing TCP CONNECT handler uses `dualStream()` for HTTP/2 (reading from `r.Body` + writing to `ResponseWriter`). The UDP handler should follow the same pattern.

**2. `ReceiveBuffer` — remote crash via malicious Length**

```go
data.Length, err = quicvarint.Read(rr)  // up to 2^62-1
bb := b[:data.Length]                    // panic if Length > len(b)
```

A malicious client sending a large Length value will panic the server. Needs bounds check:
```go
if data.Length > uint64(len(b)) {
    return fmt.Errorf("datagram too large: %d > %d", data.Length, len(b))
}
```

**3. HTTP/1.1 `Upgrade` header typo**

```go
w.Header().Set("Upgrade:", RequestProtocol)  // extra colon in header name
```

Should be `"Upgrade"`, not `"Upgrade:"`.

### Should Fix

**4. ACL logic duplicated** — The anonymous function in `tryUDPoverHTTP` duplicates `dialContextCheckACL`. Should extract a shared helper to avoid divergence.

**5. `target_port` range not validated** — `strconv.Atoi` succeeds for 0, negative, or >65535 values. RFC 9298 requires 1-65535.

**6. `ReadPacket` lacks bounds checking** — Parsing `bb[1]`, `bb[2]`...`bb[19]` without checking `len(bb)` first. Short payloads will panic.

**7. Non-standard extensions not documented** — `CompressionAssignValue (0x1C0FE323)` / `CompressionCloseValue (0x1C0FE324)` don't match current `draft-ietf-masque-connect-udp-listen` values (`0x11`/`0x12`/`0x13`), and `COMPRESSION_ACK` is missing. `Connect-UDP-Bind` is also draft-only. These should be clearly documented as experimental.

**8. `HandlePacketBind` returns error at runtime** — HTTP/3 + bind mode should be rejected at request parsing time with a proper HTTP status code, not deep in the handler.

### Minor

- All types (`Payload`, `Datagram`, `BytePayload`, etc.) and vars (`CapsuleProtocolHeaderValue`, etc.) are exported but are internal implementation details — should be unexported.
- `HandlePacket` relies on `b[0]` being zero-initialized for Context ID 0 — should explicitly set `b[0] = 0`.
- `RequestMatcher` uses `(.+)` which matches across path segments — consider `([^/]+)`.

aaravrav commented:
@imgk Is the above blocking?
